Allow @isolated(any) mismatches in witness matching.

rjmccall opened the pull request:
Fixes rdar://125394096

rjmccall commented:
@swift-ci Please test

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -5578,25 +5578,47 @@ bool ConstraintSystem::repairFailures(
+        }
+      };
+      maybeRemoveIsolatedAny(strippedLHS);
+      maybeRemoveIsolatedAny(strippedRHS);
```
This means that `isolated(any)` requirement could be witnessed by a function type without `isolated(any)` is that, okay? Seems like it might be worth to add a test for that...

rjmccall replied:
Requirements can't themselves be `@isolated(any)`, only function types.  I do have a test for a requirement that *takes* an `@isolated(any)` function, though.

xedin replied:
What I meant is - is it correct for a requirement have a function type parameter that is `isolated(any)` which is then witnessed by a corresponding function type in the witness parameter that doesn’t have that isolation? You are stripping this from both sides here and there is no new test that checks this case.

xedin replied:
As in:

```
protocol Enqueuer {
  associatedtype Result
  func enqueue(operation: @escaping @isolated(any) () async -> Result)
}

struct A<T> : Enqueuer {
  func enqueue(operation: @escaping () async -> T) {}
}
```

rjmccall replied:
I did have a test case that tested that, actually.  But I've decided not to be permissive here; it seems unnecessarily confusing (and would need extra restrictions, e.g. around `async`) and shouldn't be necessary for source compatibility anyway.

rjmccall committed:
[NFC] Paranoia and grammar

rjmccall committed:
Allow @isolated(any) mismatches in witness matching.

I'm not really convinced that this shouldn't be done by introducing a new
kind of constraint rather than hacking in what are essentially conversions
as "bind" constraints, but this is the most direct path for now.

Fixes rdar://125394096

rjmccall commented:
@swift-ci Please test

xedin reviewed (approved):
LGTM! Thank you for fixing typos!
